include/libvhd.h

```diff
@@ -165,21 +163,21 @@ struct vhd_context {
 };
 
 static inline int
-test_bit (volatile char *addr, int nr)
+test_bit (const void *addr, uint32_t nr)
 {
-	return ((addr[nr >> 3] << (nr & 7)) & BIT_MASK) != 0;
+	return (((const uint8_t *)addr)[nr >> 3] >> (nr & 7)) & 1;
 }
 
 static inline void
-set_bit (volatile char *addr, int nr)
+set_bit (void *addr, uint32_t nr)
 {
-	addr[nr >> 3] |= (BIT_MASK >> (nr & 7));
+	((uint8_t *)addr)[nr >> 3] |= (1 << (nr & 7));
```

Comment:
These two lines are not equivalent.  Consider `nr` being a multiple of 8, so `nr & 7` is 0.  In the old code you tested the top bit of the byte, and in the new code you're testing the bottom bit.

Overall, you've done a bit-reverse on the byte as it's stored in the bitmap.

AFAICT, the `old_*_bit()` were reasonably sane (on a little endian system at least), whereas the new `*_bit()` are insane on both little and bit endian systems.

Reply:
Yeah, spec says it's byte granular bit-reversed.

https://github.com/libyal/libvhdi/blob/6c102a20ad8d0c93330bc6a97beb97718befd26b/documentation/Virtual%20Hard%20Disk%20(VHD)%20image%20format.asciidoc?plain=1#L354-L355

This deserves a comment explaining the format of the bitmap.